### PR TITLE
fix(basic_types): bincode deserialization for `web3::Bytes`

### PR DESCRIPTION
**Pull request by haraldh**
## What ❔

Only use custom format, if human readable serializer is used.

## Why ❔

Don't assume we only use serde_json... Make bincode ser/deser work

## Checklist

- [x] PR title corresponds to the body of PR (we generate changelog entries from PRs).
- [ ] Tests for the changes have been added / updated.
- [ ] Documentation comments have been added / updated.
- [x] Code has been formatted via `zk fmt` and `zk lint`.
- [ ] Spellcheck has been run via `zk spellcheck`.

### CODE REVIEW

**Comment by pbeza** [comment]
You might consider pulling in my simple test for your fix: https://github.com/matter-labs/zksync-era/pull/1933 (it properly fails if you revert your fix).

**Review by popzxc — DISMISSED** [review]
LGTM, but would suggest to wait for a review from @slowli just in case.
Also, a unit test would be appreciated.

**Comment by popzxc** [comment]
@haraldh Removed from the queue due to my comment above. Feel free to re-add if you disagree :)

**Comment by haraldh** [comment]
Added the unit test from https://github.com/matter-labs/zksync-era/pull/1933 and extended it

**Review by popzxc — APPROVED** [review]

**Comment by haraldh** [comment]
ok, waiting for a review from @slowli

File: core/lib/basic_types/src/web3/mod.rs
Comment (slowli) [inline core/lib/basic_types/src/web3/mod.rs:83]: I recall there being a gotcha with this approach if the deserialized field is present in a `#[serde(flatten)]` group; I've encountered this in [my crate that implements similar functionality](https://github.com/slowli/hex-buffer-serde/blob/d203ab86d3e1772412728df19e3dca3d0eb6f447/src/var_len.rs#L305). Essentially, `serde` uses (or at least used in the past) a surrogate deserializer for such fields with `is_human_readable()` always set to true. This necessitates the original `BytesVisitor` supporting deserialization from bytes, even if it doesn't make logical sense. I guess in this case we can say that this field isn't supposed to be present in `#[serde(flatten)]` groups, but for completeness it could make sense to extend `BytesVisitor`.

Comment (popzxc) [reply core/lib/basic_types/src/web3/mod.rs:83]: IIRC `bincode` doesn't support flattening anyways.

Comment (haraldh) [reply core/lib/basic_types/src/web3/mod.rs:83]: anyways... added some `BytesVisitor` code ... plz re-review

**Review by popzxc — APPROVED** [review]

**Pull request merged**
